tasks/upload_finisher.py

```diff
@@ -325,6 +393,117 @@ def invalidate_caches(self, redis_connection, commit: Commit):
                 redis_connection.hdel("badge", ("%s:" % key).lower())
 
+    def merge_incremental_reports(
+        self,
+        commit_yaml,
+        repository,
+        commit,
+        report_code,
+        report_service,
+        processing_results,
+        db_session,
+    ):
+        archive_service = report_service.get_archive_service(repository)
+        repoid = repository.repoid
+        commitid = commit.id
+
+        fas_path = MinioEndpoints.parallel_upload_experiment.get_path(
+            version="v4",
+            repo_hash=archive_service.get_archive_hash(repository),
+            commitid=commit.commitid,
+            file_name="files_and_sessions",
+        )
+        chunks_path = MinioEndpoints.parallel_upload_experiment.get_path(
+            version="v4",
+            repo_hash=archive_service.get_archive_hash(repository),
+            commitid=commit.commitid,
+            file_name="chunks",
+        )
+
+        try:
+            files_and_sessions = json.loads(archive_service.read_file(fas_path))
+            chunks = archive_service.read_file(chunks_path).decode(errors="replace")
+            report = report_service.build_report(
+                chunks,
+                files_and_sessions["files"],
+                files_and_sessions["sessions"],
+                None,
+            )
+        except FileNotInStorageError:  # there were no CFFs, so no report was stored in GCS
+            report = Report()
+
+        log.info(
+            "Downloading %s incremental reports that were processed in parallel",
+            len(processing_results["processings_so_far"]),
+            extra=dict(
+                repoid=repoid,
+                commit=commitid,
+                processing_results=processing_results["processings_so_far"],
+                parent_task=self.request.parent_id,
+            ),
+        )
+
+        def download_and_build_incremental_report(partial_report):
+            chunks = archive_service.read_file(partial_report["chunks_path"]).decode(
+                errors="replace"
+            )
+            files_and_sessions = json.loads(
+                archive_service.read_file(partial_report["files_and_sessions_path"])
+            )
+            report = report_service.build_report(
+                chunks,
+                files_and_sessions["files"],
+                files_and_sessions["sessions"],
+                None,
+            )
+            return {
+                "parallel_idx": partial_report["parallel_idx"],
+                "report": report,
+                "upload_pk": partial_report["upload_pk"],
+            }
+
+        def merge_report(cumulative_report, obj):
+            incremental_report = obj["report"]
+            parallel_idx = obj["parallel_idx"]
+            assert len(incremental_report.sessions) == 1
+            session_id, session = cumulative_report.add_session(
+                incremental_report.sessions[parallel_idx]
+            )
```

Comment:
shouldn't this use `use_id_from_session`? otherwise if you insert something with `parallel_idx=15` before something with `parallel_idx=12` it'll get mixed up?

Reply:
good catch... I'm confused as to why this didn't raise problems earlier...